File: src/main/java/com/github/hirsivaja/ip/ipv4/Ipv4Payload.java
package com.github.hirsivaja.ip.ipv4;

import com.github.hirsivaja.ip.*;
import com.github.hirsivaja.ip.icmp.IcmpPayload;
import com.github.hirsivaja.ip.igmp.IgmpPayload;
import com.github.hirsivaja.ip.tcp.TcpSegment;
import com.github.hirsivaja.ip.udp.UdpDatagram;

import java.nio.ByteBuffer;

public sealed interface Ipv4Payload extends IpPayload permits IcmpPayload, IgmpPayload, AuthenticationPayload, EncapsulationPayload, EspPayload, Ipv4Payload.Generic, TcpSegment, UdpDatagram {

    static Ipv4Payload decode(ByteBuffer in, boolean ensureChecksum, Ipv4Header header) {
        return switch (header.protocol()) {
            case IpProtocols.ICMP -> IcmpPayload.decode(in, ensureChecksum);
            case IpProtocols.IGMP -> IgmpPayload.decode(in);
            case IpProtocols.TCP -> TcpSegment.decode(in, ensureChecksum, header);
            case IpProtocols.UDP -> UdpDatagram.decode(in, ensureChecksum, header);
            case IpProtocols.IPV6_ENCAPSULATION -> EncapsulationPayload.decode(in, ensureChecksum);
            case IpProtocols.ESP -> EspPayload.decode(in);
            case IpProtocols.AUTHENTICATION -> AuthenticationPayload.decode(in, ensureChecksum);
            default -> Ipv4Payload.Generic.decode(in, header.protocol());
        };
    }

    record Generic(IpProtocol protocol, ByteArray data) implements Ipv4Payload {

        @Override
        public void encode(ByteBuffer out) {
            out.put(data.array());
        }

        @Override
        public int length() {
            return data.array().length;
        }

        public static Ipv4Payload decode(ByteBuffer in, IpProtocol protocol) {
            byte[] data = new byte[in.remaining()];
            in.get(data);
            return new Generic(protocol, new ByteArray(data));
        }
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/payload/Ipv4Payload.kt
package com.github.hirsivaja.ip.packet.payload

import com.github.hirsivaja.ip.DataArray
import com.github.hirsivaja.ip.packet.header.Ipv4Header
import com.github.hirsivaja.ip.packet.payload.tcp.TcpSegment
import com.github.hirsivaja.ip.packet.payload.udp.UdpDatagram
import com.github.hirsivaja.ip.protocol.IpProtocol
import com.github.hirsivaja.ip.protocol.IpProtocols
import java.nio.ByteBuffer

interface Ipv4Payload : IpPayload {
    companion object {
        fun decode(buf: ByteBuffer, ensureChecksum: Boolean, header: Ipv4Header) : Ipv4Payload {
            return when (header.protocol) {
                IpProtocols.TCP -> TcpSegment.decode(buf, ensureChecksum, header)
                IpProtocols.UDP -> UdpDatagram.decode(buf, ensureChecksum, header)
                else -> Generic.decode(buf, header.protocol)
            } as Ipv4Payload
        }
    }

    data class Generic(val protocol: IpProtocol, val data: DataArray) : Ipv4Payload {
        constructor(protocol: IpProtocol, data: ByteArray) : this(protocol, DataArray(data))

        override fun encode(out: ByteBuffer) {
            out.put(data.data)
        }

        override fun length(): Int {
            return data.length()
        }

        companion object {
            fun decode(buf: ByteBuffer, protocol: IpProtocol): IpPayload {
                val data = ByteArray(buf.remaining())
                buf[data]
                return Generic(protocol, data)
            }
        }
    }
}
